docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,319 @@
+this condition can exist is the lease duration.
+
+# Drawbacks
+
+* This is complex. Perhaps overly complex. Where can it be simplified?
+
+* The watch mechanism is described with the assumption of streaming
+  responses. Unless we adopt gRPC sooner, we'll have to fallback to
+  some sort of cursor mechanism.
+
+# Alternatives
+
+* We could use an existing lock service such as etcd or Zookeeper. The
+  primary benefit would be the builtin watch functionality. We would
+  still need the logic for local reference counts.
+
+* The lease master could avoid storing the lease state, but we would
+  have to wait for the lease duration at master startup before
+  allowing any schema modifications. There might be some hacks around
+  this. For example, The master can persist only the largest lease
+  expiration and wait until that expiration passes before allowing
+  schema modifications. On the other hand, persisting the lease state
+  allows inspection of it via the SQL console.
+
+* Keeping track of local references to descriptor versions in order to
+  early release leases adds complexity. We could just wait for leases
+  to expire, though that would cause a 3-step schema modification to
+  take at least 10m to complete.
+
+* We could avoid having a lease master. Nodes would add/remove leases
+  directly to/from the `leaseTable`. Publish would require some sort
+  of conditional put on the new version. Watching would be replaced by
+  polling. While this seems feasible, it also seems quite a bit more
+  subtle than a centralized coordinator for leases for a table.
+
+* We could continue to distribute descriptors using gossip and add
+  some sort of high-priority flag to indicate that we want to skip the
+  current 2s/hop delay in gossip. This would replace the watch
+  mechanism for discovery of new descriptors, though we would still
+  need a watch mechanism for discovery of `DEAD` descriptors.
+
+# Unresolved questions
+
+* We need a mechanism to enforce that write operations do not commit
+  if they reference a table descriptor lease which has expired. The F1
+  paper talks about a "write fence" mechanism in spanner which allows
+  a transaction to specify a time before which the transaction needs
+  to commit otherwise it should be aborted.
+
+* Is `leaseDuration/2` the right minimum time to give an operation for
+  usage of a lease? Note that the maximum time is unbounded if the
+  node can successfully extend the lease duration during the
+  operation. If the lease can't be extended (e.g. because of a
+  concurrent schema modification) then the operation will need to be
+  aborted. So the question becomes how common are operations that will
+  take longer than `leaseDuration/2` and how problematic is it if they
+  are aborted?
+
+* How to direct RPCs to the leader for a particular range? The table
+  leasing code is SQL specific, but all existing special handling for
+  leader operations happens in `storage`.
```

Comment:
We just need to treat these RPCs as "admin" operations (as opposed to read or write). See the handling of AdminSplit and AdminMerge in `Replica.addAdminCommand`. This will return a `NotLeaderError` if they reach a follower and then the caller can retry.